### ISSUE TITLE
Unable to query `Query Docs` mode when using sagemaker

### ISSUE THREAD

**Issue by LvffY** [imartinez/privateGPT#1367]
## Context

Hi everyone, 

I have send [a message through Discord](https://discord.com/channels/1164200432894234644/1164200433779212400/1181551995870523423) but thought it would be easier to manage the issue here.

What I'm trying to achieve is to run privateGPT with some production-grade environment. To do so, I've tried to run something like : 

- Create a Qdrant database in Qdrant cloud
- Run LLM model and embedding model through Sagemaker

For I have the following setups : 

| LLM  | Embedding | Qdrant |  State
| ------------- | ------------- | ------------- | ------------- | 
| Local | Local | Local | Success |
| Local | Local | Qdrant cloud | Success |
| Sagemaker | Local | Local | Failed |
| Sagemaker | Sagemaker | Qdrant cloud | Failed |

As we can see, whenever I'm trying to use sagemaker it seems to fail to use the RAG architecture.

## How to reproduce

I've tried to have the simplest setup to reproduce, if you want me to test anything else, do not hesitate to ask me.

- Create a new profile **sagemaker** with **settings-sagemaker.yaml** : 

```yaml
server:
  env_name: ${APP_ENV:prod}
  port: ${PORT:8001}

ui:
  enabled: true
  path: /

llm:
  mode: sagemaker

embedding:
  # Should be matching the value above in most cases
  mode: local
  ingest_mode: simple

sagemaker:
  llm_endpoint_name: TheBloke-Mistral-7B-Instruct-v0-1-GPTQ # This should be first deployed in your sagemaker instance
  embedding_endpoint_name: BAAI-bge-large-en-v1-5 # This should be first deployed in your sagemaker instance
```
- Ingest a file through the UI
- Run `PGPT_PROFILES=sagemaker make run`
- Ask a question related to your file. (Weirdly, if you ask a question completly unrelated it may work ...)
 
## Actual behavior

- The UI is sending a weird output
![image](https://github.com/imartinez/privateGPT/assets/37327656/a25a8f65-bd6c-4abc-89a1-eec085412b26)
- We see some weird warning related to **llama_index**
![image](https://github.com/imartinez/privateGPT/assets/37327656/3aca095c-22fb-467e-a05f-0ae12454c203)

## Expected behavior

To be able to use the `Query Docs` mode even when using sagemaker.

**Comment by LvffY** [comment]
Thanks @logan-markewich I'll try to dig further on this way.

As I'm not an expert, this could take a while, so if anyone has a solution or want to try, be my guest :)

Also, I've tried with the latest **0.2.0** release and I see the same behavior.

**Comment by LvffY** [comment]
> Probably this line of code here
> 
> https://github.com/imartinez/privateGPT/blob/9302620eaca56d00818cb4db87ea1e8a8aa170f9/private_gpt/components/llm/custom/sagemaker.py#L256

Based on this message, I've dig a little bit more and it seems that the prompt message is not passed to the endpoint. 

And then because the prompt is empty, it seems that the endpoint sends an uncatch exception in the message (but the endpoint clearly send back a 200 api call, so this is not really an exception here.)

It seems that's this line that returns an empty prompt.

https://github.com/imartinez/privateGPT/blob/e8ac51bba4b698c8a66dfd02bda5020f4a08f0cd/private_gpt/components/llm/custom/sagemaker.py#L273

I've tried to add some prints and I got this output : 

![image](https://github.com/imartinez/privateGPT/assets/37327656/2b53b771-a08a-4e6a-bf5b-3d8ac0fca59b)

I'm not used to the `Field` object so don't know where to start. If you have any idea @logan-markewich (or @pabloogc because it's you that started this code).

I'll try to run more tests, don't hesitate if you have questions

**Comment by LvffY** [comment]
I think I have a solution to my problem. Thanks to the help of @logan-markewich (in [this discord thread](https://discord.com/channels/1059199217496772688/1182591715136262144)) I've dig quite a bit through the code and it appears to be an issue with the memory of the chat.

The "cause" was the retrieval of quite a long context that would end up in the LLM just receive the context and not my question.

I fixed it with two different solutions (and keep only the last one) : 

- You can change the return **_chat_engine** method from the **chat_service** module, by adding `ChatMemoryBuffer.from_defaults(token_limit=3900)` to increase the token limit. (Of course, *3900* is arbitrary)

- Digging a bit more, and looking into the very specific [ChatMemoryBuffer.from_defaults method](https://github.com/run-llama/llama_index/blob/f9a14dff27c87f21c05e484666195e3f2a60aea3/llama_index/memory/chat_memory_buffer.py#L50), it appeared to me that this limit could also be manipulated through the **context_window** parameter and this **context_window** parameter was already configured for the [LLamaCPP LLM](https://github.com/imartinez/privateGPT/blob/3582764801c692f53143a2f895c87005f845a4b7/private_gpt/components/llm/llm_component.py#L34). Hence it seems that extracting this parameter from the settings could be a good idea. Based on that : 

  - Update the [LLMSettings](https://github.com/imartinez/privateGPT/blob/3582764801c692f53143a2f895c87005f845a4b7/private_gpt/settings/settings.py#L83) class to add a new parameter.
  -  Update the [LLMComponent](https://github.com/imartinez/privateGPT/blob/3582764801c692f53143a2f895c87005f845a4b7/private_gpt/components/llm/llm_component.py#L15) class to add a new parameter.

```python
class LLMSettings(BaseModel):
    mode: Literal["local", "openai", "sagemaker", "mock"]
    max_new_tokens: int = Field(
        256,
        description="The maximum number of token that the LLM is authorized to generate in one completion.",
    )
    ## This is new
    context_window_size: int = Field(
        3900,
        description=(
            "Size of the context window.\n"
            "llama2 has a context window of 4096 tokens, but we set it lower to allow for some wiggle room.\n"
            "You may need to increase this context window, see https://github.com/imartinez/privateGPT/issues/1367."
        ),
    )
```

```python
            case "local":
                from llama_index.llms import LlamaCPP

                prompt_style = get_prompt_style(settings.local.prompt_style)

                self.llm = LlamaCPP(
                    model_path=str(models_path / settings.local.llm_hf_model_file),
                    temperature=0.1,
                    max_new_tokens=settings.llm.max_new_tokens,
                    context_window=settings.llm.context_window_size, # This line is changed
                    generate_kwargs={},
                    # All to GPU
                    model_kwargs={"n_gpu_layers": -1},
                    # transform inputs into Llama2 format
                    messages_to_prompt=prompt_style.messages_to_prompt,
                    completion_to_prompt=prompt_style.completion_to_prompt,
                    verbose=True,
                )

            case "sagemaker":
                from private_gpt.components.llm.custom.sagemaker import SagemakerLLM

                self.llm = SagemakerLLM(
                    endpoint_name=settings.sagemaker.llm_endpoint_name,
                    max_new_tokens=settings.llm.max_new_tokens,
                    context_window=settings.llm.context_window_size, # This line is added
                )
```

If you think it's useful, I could make some PR

**Comment by imartinez** [comment]
I'm working on a PR that makes context_window and max_new_tokens customizable in settings.yaml

Using the right context_window is the way to go imo. 

Thanks for the support and the documentation!

**Comment by LvffY** [comment]
Closed by #1437
